File: library/src/main/java/org/xmtp/android/library/libxmtp/DecodedMessageV2.kt
package org.xmtp.android.library.libxmtp

import android.util.Log
import com.google.protobuf.kotlin.toByteString
import org.xmtp.android.library.InboxId
import org.xmtp.android.library.codecs.Attachment
import org.xmtp.android.library.codecs.ContentTypeId
import org.xmtp.android.library.codecs.ContentTypeIdBuilder
import org.xmtp.android.library.codecs.MultiRemoteAttachment
import org.xmtp.android.library.codecs.Reaction
import org.xmtp.android.library.codecs.ReactionAction
import org.xmtp.android.library.codecs.ReactionSchema
import org.xmtp.android.library.codecs.ReadReceipt
import org.xmtp.android.library.codecs.RemoteAttachment
import org.xmtp.android.library.codecs.RemoteAttachmentInfo
import org.xmtp.android.library.codecs.TransactionReference
import org.xmtp.android.library.codecs.decoded
import org.xmtp.android.library.codecs.encodedContentFromFfi
import org.xmtp.android.library.toHex
import uniffi.xmtpv3.FfiAttachment
import uniffi.xmtpv3.FfiDecodedMessage
import uniffi.xmtpv3.FfiDecodedMessageBody
import uniffi.xmtpv3.FfiDecodedMessageContent
import uniffi.xmtpv3.FfiDeliveryStatus
import uniffi.xmtpv3.FfiGroupUpdated
import uniffi.xmtpv3.FfiInbox
import uniffi.xmtpv3.FfiMetadataFieldChange
import uniffi.xmtpv3.FfiMultiRemoteAttachment
import uniffi.xmtpv3.FfiReactionAction
import uniffi.xmtpv3.FfiReactionPayload
import uniffi.xmtpv3.FfiReactionSchema
import uniffi.xmtpv3.FfiRemoteAttachment
import uniffi.xmtpv3.FfiRemoteAttachmentInfo
import uniffi.xmtpv3.FfiTransactionMetadata
import uniffi.xmtpv3.FfiTransactionReference
import java.net.URL
import java.util.Date

class DecodedMessageV2 private constructor(
    private val libXMTPMessage: FfiDecodedMessage,
) {
    val id: String
        get() = libXMTPMessage.id().toHex()

    val conversationId: String
        get() = libXMTPMessage.conversationId().toHex()

    val senderInboxId: InboxId
        get() = libXMTPMessage.senderInboxId()

    val sentAt: Date
        get() = Date(libXMTPMessage.sentAtNs() / 1_000_000)

    val sentAtNs: Long
        get() = libXMTPMessage.sentAtNs()

    val insertedAtNs: Long
        get() = libXMTPMessage.insertedAtNs()

    val deliveryStatus: DecodedMessage.MessageDeliveryStatus
        get() =
            when (libXMTPMessage.deliveryStatus()) {
                FfiDeliveryStatus.UNPUBLISHED -> DecodedMessage.MessageDeliveryStatus.UNPUBLISHED
                FfiDeliveryStatus.PUBLISHED -> DecodedMessage.MessageDeliveryStatus.PUBLISHED
                FfiDeliveryStatus.FAILED -> DecodedMessage.MessageDeliveryStatus.FAILED
            }

    val reactions: List<DecodedMessageV2>
        get() = libXMTPMessage.reactions().mapNotNull { create(it) }

    val hasReactions: Boolean
        get() = libXMTPMessage.hasReactions()

    val reactionCount: ULong
        get() = libXMTPMessage.reactionCount()

    val fallbackText: String?
        get() = libXMTPMessage.fallbackText()

    val contentTypeId: ContentTypeId
        get() = ContentTypeIdBuilder.fromFfi(libXMTPMessage.contentTypeId())

    @Suppress("UNCHECKED_CAST")
    fun <T> content(): T? =
        try {
            decodeContent(libXMTPMessage.content()) as? T
        } catch (e: Exception) {
            Log.e("DecodedMessageV2", "Error decoding content: ${e.message}")
            null
        }

    companion object {
        fun create(libXMTPMessage: FfiDecodedMessage): DecodedMessageV2? =
            try {
                DecodedMessageV2(libXMTPMessage)
            } catch (e: Exception) {
                Log.e("DecodedMessageV2", "Error creating DecodedMessageV2: ${e.message}")
                null
            }

        // Helper functions for mapping FFI types to domain types

        private fun mapReaction(ffiReaction: FfiReactionPayload): Reaction {
            val action =
                when (ffiReaction.action) {
                    FfiReactionAction.ADDED -> ReactionAction.Added
                    FfiReactionAction.REMOVED -> ReactionAction.Removed
                    FfiReactionAction.UNKNOWN -> ReactionAction.Unknown
                }
            val schema =
                when (ffiReaction.schema) {
                    FfiReactionSchema.UNICODE -> ReactionSchema.Unicode
                    FfiReactionSchema.SHORTCODE -> ReactionSchema.Shortcode
                    FfiReactionSchema.CUSTOM -> ReactionSchema.Custom
                    FfiReactionSchema.UNKNOWN -> ReactionSchema.Unknown
                }
            return Reaction(
                reference = ffiReaction.reference,
                action = action,
                content = ffiReaction.content,
                schema = schema,
            )
        }

        private fun mapAttachment(ffiAttachment: FfiAttachment): Attachment =
            Attachment(
                filename = ffiAttachment.filename ?: "",
                mimeType = ffiAttachment.mimeType,
                data = ffiAttachment.content.toByteString(),
            )

        private fun mapRemoteAttachment(ffiRemote: FfiRemoteAttachment): RemoteAttachment =
            RemoteAttachment(
                url = URL(ffiRemote.url),
                contentDigest = ffiRemote.contentDigest,
                secret = ffiRemote.secret.toByteString(),
                salt = ffiRemote.salt.toByteString(),
                nonce = ffiRemote.nonce.toByteString(),
                scheme = ffiRemote.scheme,
                contentLength = ffiRemote.contentLength.toInt(),
                filename = ffiRemote.filename,
            )

        private fun mapRemoteAttachmentInfo(ffiInfo: FfiRemoteAttachmentInfo): RemoteAttachmentInfo =
            RemoteAttachmentInfo(
                url = ffiInfo.url,
                filename = ffiInfo.filename ?: "",
                contentLength = ffiInfo.contentLength?.toLong() ?: 0,
                contentDigest = ffiInfo.contentDigest,
                nonce = ffiInfo.nonce.toByteString(),
                scheme = ffiInfo.scheme,
                salt = ffiInfo.salt.toByteString(),
                secret = ffiInfo.secret.toByteString(),
            )

        private fun mapMultiRemoteAttachment(ffiMulti: FfiMultiRemoteAttachment): MultiRemoteAttachment =
            MultiRemoteAttachment(
                remoteAttachments = ffiMulti.attachments.map { mapRemoteAttachmentInfo(it) },
            )

        private fun mapTransactionMetadata(meta: FfiTransactionMetadata): TransactionReference.Metadata =
            TransactionReference.Metadata(
                transactionType = meta.transactionType,
                currency = meta.currency,
                amount = meta.amount,
                decimals = meta.decimals,
                fromAddress = meta.fromAddress,
                toAddress = meta.toAddress,
            )

        private fun mapTransactionReference(ffiTx: FfiTransactionReference): TransactionReference =
            TransactionReference(
                namespace = ffiTx.namespace,
                networkId = ffiTx.networkId,
                reference = ffiTx.reference,
                metadata = ffiTx.metadata?.let { mapTransactionMetadata(it) },
            )

        // Helper functions for GroupUpdated proto mapping

        private fun mapFfiInboxToProto(
            ffiInbox: FfiInbox,
        ): org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated.Inbox =
            org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated.Inbox
                .newBuilder()
                .setInboxId(ffiInbox.inboxId)
                .build()

        private fun mapFfiMetadataChangeToProto(
            ffiChange: FfiMetadataFieldChange,
        ): org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated.MetadataFieldChange =
            org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated.MetadataFieldChange
                .newBuilder()
                .setFieldName(ffiChange.fieldName)
                .apply {
                    ffiChange.oldValue?.let { setOldValue(it) }
                    ffiChange.newValue?.let { setNewValue(it) }
                }.build()

        private fun mapGroupUpdated(
            ffiGroupUpdated: FfiGroupUpdated,
        ): org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated =
            org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated
                .newBuilder()
                .apply {
                    initiatedByInboxId = ffiGroupUpdated.initiatedByInboxId

                    // Add added inboxes
                    ffiGroupUpdated.addedInboxes.forEach { ffiInbox ->
                        addAddedInboxes(mapFfiInboxToProto(ffiInbox))
                    }

                    // Add removed inboxes
                    ffiGroupUpdated.removedInboxes.forEach { ffiInbox ->
                        addRemovedInboxes(mapFfiInboxToProto(ffiInbox))
                    }

                    // Add metadata field changes
                    ffiGroupUpdated.metadataFieldChanges.forEach { ffiChange ->
                        addMetadataFieldChanges(mapFfiMetadataChangeToProto(ffiChange))
                    }
                }.build()

        /**
         * Decode content from FfiDecodedMessageContent
         */
        internal fun decodeContent(content: FfiDecodedMessageContent): Any? =
            when (content) {
                is FfiDecodedMessageContent.Text -> content.v1.content
                is FfiDecodedMessageContent.Reaction -> mapReaction(content.v1)
                is FfiDecodedMessageContent.Reply -> Reply.create(content.v1)
                is FfiDecodedMessageContent.Attachment -> mapAttachment(content.v1)
                is FfiDecodedMessageContent.RemoteAttachment -> mapRemoteAttachment(content.v1)
                is FfiDecodedMessageContent.MultiRemoteAttachment ->
                    mapMultiRemoteAttachment(
                        content.v1,
                    )

                is FfiDecodedMessageContent.TransactionReference -> mapTransactionReference(content.v1)
                is FfiDecodedMessageContent.WalletSendCalls -> content.v1
                is FfiDecodedMessageContent.GroupUpdated -> mapGroupUpdated(content.v1)
                is FfiDecodedMessageContent.ReadReceipt -> ReadReceipt
                is FfiDecodedMessageContent.Custom -> {
                    val encodedContent = encodedContentFromFfi(content.v1)
                    encodedContent.decoded<Any>()
                }

                else -> null
            }

        /**
         * Decode content from FfiDecodedMessageBody (used by Reply)
         */
        internal fun decodeBodyContent(body: FfiDecodedMessageBody): Any? =
            when (body) {
                is FfiDecodedMessageBody.Text -> body.v1.content
                is FfiDecodedMessageBody.Reaction -> mapReaction(body.v1)
                is FfiDecodedMessageBody.Attachment -> mapAttachment(body.v1)
                is FfiDecodedMessageBody.RemoteAttachment -> mapRemoteAttachment(body.v1)
                is FfiDecodedMessageBody.MultiRemoteAttachment -> mapMultiRemoteAttachment(body.v1)
                is FfiDecodedMessageBody.TransactionReference -> mapTransactionReference(body.v1)
                is FfiDecodedMessageBody.WalletSendCalls -> body.v1
                is FfiDecodedMessageBody.GroupUpdated -> mapGroupUpdated(body.v1)
                is FfiDecodedMessageBody.Custom -> {
                    val encodedContent = encodedContentFromFfi(body.v1)
                    encodedContent.decoded<Any>()
                }

                else -> null
            }
    }
}

File: sdks/ios/Sources/XMTPiOS/Libxmtp/DecodedMessageV2.swift
import Foundation

/// A type representing a structured intent from a message.
public typealias Intent = FfiIntent
/// A type representing structured actions from a message.
public typealias Actions = FfiActions

/// An enriched decoded message with reactions and replies baked in.
///
/// Unlike ``DecodedMessage``, `DecodedMessageV2` includes enriched metadata such as
/// reactions and the original message for replies directly on the message object.
/// This eliminates the need for separate queries and is recommended for UI rendering.
///
/// Retrieve enriched messages via ``Group/enrichedMessages(beforeNs:afterNs:limit:direction:deliveryStatus:excludeContentTypes:excludeSenderInboxIds:sortBy:insertedAfterNs:insertedBeforeNs:)``
/// or ``Dm/enrichedMessages(beforeNs:afterNs:limit:direction:deliveryStatus:excludeContentTypes:excludeSenderInboxIds:sortBy:insertedAfterNs:insertedBeforeNs:)``.
///
/// ```swift
/// let messages = try await group.enrichedMessages(limit: 20)
/// for message in messages {
///     let text: String = try message.content()
///     if let reactions = message.reactions {
///         // Handle reactions
///     }
/// }
/// ```
public struct DecodedMessageV2: Identifiable {
	private let ffiMessage: FfiDecodedMessage

	/// The hex-encoded unique identifier of this message.
	public var id: String {
		ffiMessage.id().toHex
	}

	/// The hex-encoded identifier of the conversation this message belongs to.
	public var conversationId: String {
		ffiMessage.conversationId().toHex
	}

	/// The inbox ID of the account that sent this message.
	public var senderInboxId: InboxId {
		ffiMessage.senderInboxId()
	}

	/// The date when this message was sent on the network.
	public var sentAt: Date {
		Date(timeIntervalSince1970: TimeInterval(ffiMessage.sentAtNs()) / 1_000_000_000)
	}

	/// The timestamp in nanoseconds when this message was sent on the network.
	public var sentAtNs: Int64 {
		ffiMessage.sentAtNs()
	}

	/// The date when this message was inserted into the local database.
	public var insertedAt: Date {
		Date(timeIntervalSince1970: TimeInterval(ffiMessage.insertedAtNs()) / 1_000_000_000)
	}

	/// The timestamp in nanoseconds when this message was inserted into the local database.
	public var insertedAtNs: Int64 {
		ffiMessage.insertedAtNs()
	}

	/// The timestamp in nanoseconds when this message expires, or `nil` if it does not expire.
	public var expiresAtNs: Int64? {
		ffiMessage.expiresAtNs()
	}

	/// The date when this message expires, or `nil` if it does not expire.
	public var expiresAt: Date? {
		expiresAtNs.map { Date(timeIntervalSince1970: TimeInterval($0) / 1_000_000_000) }
	}

	/// The current delivery status of this message.
	public var deliveryStatus: MessageDeliveryStatus {
		switch ffiMessage.deliveryStatus() {
		case .unpublished:
			.unpublished
		case .published:
			.published
		case .failed:
			.failed
		}
	}

	/// The MLS topic string for the conversation this message belongs to.
	public var topic: String {
		Topic.groupMessage(conversationId).description
	}

	/// The reactions associated with this message, or `nil` if there are none.
	///
	/// Each reaction is itself a `DecodedMessageV2` whose content can be decoded
	/// as a `Reaction` type.
	public var reactions: [DecodedMessageV2]? {
		let reactionMessages = ffiMessage.reactions()
		guard !reactionMessages.isEmpty else { return nil }
		return reactionMessages.compactMap { DecodedMessageV2(ffiMessage: $0) }
	}

	/// Extracts the decoded content of this message as the specified type.
	///
	/// The content is decoded on each call from the enriched FFI representation.
	/// For plain text messages, use `String`. For reactions, use `Reaction`, etc.
	///
	/// ```swift
	/// let text: String = try message.content()
	/// ```
	///
	/// - Returns: The decoded content cast to the requested type.
	/// - Throws: ``DecodedMessageError/decodeError(_:)`` if the content cannot be cast to type `T`.
	public func content<T>() throws -> T {
		let decodedContent = try mapContent(ffiMessage.content())
		guard let result = decodedContent as? T else {
			throw DecodedMessageError.decodeError(
				"Decoded content could not be cast to the expected type \(T.self)."
			)
		}
		return result
	}

	/// The fallback text representation of this message's content.
	///
	/// Returns the codec-provided fallback text if available, otherwise
	/// attempts to extract a human-readable string from the content.
	public var fallback: String {
		get throws {
			if let fallbackText = ffiMessage.fallbackText(), !fallbackText.isEmpty {
				return fallbackText
			}
			switch ffiMessage.content() {
			case let .text(text):
				return text.content
			case let .custom(encodedContent):
				return encodedContent.fallback ?? ""
			case .leaveRequest:
				return "A member has requested leaving the group"
			default:
				return ""
			}
		}
	}

	/// A plain-text representation of this message.
	///
	/// Returns the decoded `String` content if available, otherwise falls back
	/// to the fallback text provided by the codec.
	public var body: String {
		get throws {
			do {
				return try content() as String
			} catch {
				return try fallback
			}
		}
	}

	/// The content type identifier for this message (e.g., text, reaction, attachment).
	public var contentTypeId: ContentTypeID {
		let ffiContentType = ffiMessage.contentTypeId()
		return ContentTypeID(
			authorityID: ffiContentType.authorityId,
			typeID: ffiContentType.typeId,
			versionMajor: Int(ffiContentType.versionMajor),
			versionMinor: Int(ffiContentType.versionMinor)
		)
	}

	/// Creates a `DecodedMessageV2` from an FFI decoded message, or returns `nil` if the message is invalid.
	public init?(ffiMessage: FfiDecodedMessage) {
		self.ffiMessage = ffiMessage
	}

	/// Creates a `DecodedMessageV2` from an FFI decoded message, or returns `nil` if the message is invalid.
	public static func create(ffiMessage: FfiDecodedMessage) -> DecodedMessageV2? {
		DecodedMessageV2(ffiMessage: ffiMessage)
	}

	private func mapContent(_ content: FfiDecodedMessageContent) throws -> Any {
		switch content {
		case let .text(textContent):
			return textContent.content

		case let .markdown(markdownContent):
			return markdownContent.content

		case let .reply(enrichedReply):
			return try mapReply(enrichedReply)

		case let .reaction(reactionPayload):
			return mapReaction(reactionPayload)

		case let .attachment(ffiAttachment):
			return mapAttachment(ffiAttachment)

		case let .remoteAttachment(ffiAttachment):
			return try mapRemoteAttachment(ffiAttachment)

		case let .multiRemoteAttachment(multiAttachment):
			return mapMultiRemoteAttachment(multiAttachment)

		case let .transactionReference(txRef):
			return mapTransactionReference(txRef)

		case let .groupUpdated(groupUpdated):
			return mapGroupUpdated(groupUpdated)

		case .readReceipt:
			return ReadReceipt()

		case let .leaveRequest(ffiLeaveRequest):
			return mapLeaveRequest(ffiLeaveRequest)

		case let .walletSendCalls(walletSend):
			return walletSend

		case let .custom(ffiEncodedContent):
			let encoded = try mapFfiEncodedContent(ffiEncodedContent)
			let codec = Client.codecRegistry.find(for: encoded.type)
			return try codec.decode(content: encoded)

		case let .intent(intent):
			return intent

		case let .actions(actions):
			return actions

		case let .deletedMessage(ffiDeletedMessage):
			return mapDeletedMessage(ffiDeletedMessage)
		}
	}

	private func mapDeletedMessage(_ ffiDeletedMessage: FfiDeletedMessage) -> DeletedMessage {
		let deletedBy: DeletedBy = switch ffiDeletedMessage.deletedBy {
		case .sender:
			.sender
		case let .admin(inboxId):
			.admin(inboxId: inboxId)
		}
		return DeletedMessage(deletedBy: deletedBy)
	}

	private func mapLeaveRequest(_ ffiLeaveRequest: FfiLeaveRequest) -> LeaveRequest {
		LeaveRequest(authenticatedNote: ffiLeaveRequest.authenticatedNote)
	}

	private func mapReply(_ enrichedReply: FfiEnrichedReply) throws -> Reply {
		guard let bodyContent = enrichedReply.content else {
			throw DecodedMessageError.decodeError("Missing reply content")
		}

		let content = try mapMessageBody(bodyContent)
		let contentType = determineContentType(from: bodyContent)

		var reply = Reply(
			reference: enrichedReply.inReplyTo?.id().toHex ?? "",
			content: content,
			contentType: contentType
		)

		if let inReplyToMessage = enrichedReply.inReplyTo {
			reply.inReplyTo = DecodedMessageV2(ffiMessage: inReplyToMessage)
		}

		return reply
	}

	private func mapMessageBody(_ body: FfiDecodedMessageBody) throws -> Any {
		switch body {
		case let .text(textContent):
			return textContent.content
		case let .markdown(markdownContent):
			return markdownContent.content
		case let .reaction(reactionPayload):
			return mapReaction(reactionPayload)
		case let .attachment(ffiAttachment):
			return mapAttachment(ffiAttachment)
		case let .remoteAttachment(ffiAttachment):
			return try mapRemoteAttachment(ffiAttachment)
		case let .multiRemoteAttachment(multiAttachment):
			return mapMultiRemoteAttachment(multiAttachment)
		case let .transactionReference(txRef):
			return mapTransactionReference(txRef)
		case let .groupUpdated(groupUpdated):
			return mapGroupUpdated(groupUpdated)
		case let .walletSendCalls(walletSend):
			return walletSend
		case .readReceipt:
			return ReadReceipt()
		case let .leaveRequest(ffiLeaveRequest):
			return mapLeaveRequest(ffiLeaveRequest)
		case let .custom(ffiEncodedContent):
			let encoded = try mapFfiEncodedContent(ffiEncodedContent)
			let codec = Client.codecRegistry.find(for: encoded.type)
			return try codec.decode(content: encoded)
		case let .intent(intent):
			return intent as Intent
		case let .actions(actions):
			return actions as Actions
		case let .deletedMessage(ffiDeletedMessage):
			return mapDeletedMessage(ffiDeletedMessage)
		}
	}

	private func determineContentType(from body: FfiDecodedMessageBody) -> ContentTypeID {
		switch body {
		case .text:
			ContentTypeText
		case .markdown:
			ContentTypeID(
				authorityID: "xmtp.org",
				typeID: "markdown",
				versionMajor: 1,
				versionMinor: 0
			)
		case .reaction:
			ContentTypeReaction
		case .attachment:
			ContentTypeAttachment
		case .remoteAttachment:
			ContentTypeRemoteAttachment
		case .multiRemoteAttachment:
			ContentTypeMultiRemoteAttachment
		case .transactionReference:
			ContentTypeTransactionReference
		case .groupUpdated:
			ContentTypeGroupUpdated
		case .readReceipt:
			ContentTypeReadReceipt
		case .leaveRequest:
			ContentTypeLeaveRequest
		case .walletSendCalls:
			ContentTypeID(
				authorityID: "xmtp.org",
				typeID: "walletSendCalls",
				versionMajor: 1,
				versionMinor: 0
			)
		case let .custom(ffiEncodedContent):
			if let typeId = ffiEncodedContent.typeId {
				ContentTypeID(
					authorityID: typeId.authorityId,
					typeID: typeId.typeId,
					versionMajor: Int(typeId.versionMajor),
					versionMinor: Int(typeId.versionMinor)
				)
			} else {
				// Return a default content type if none is specified
				ContentTypeText
			}
		case .intent:
			ContentTypeID(
				authorityID: "coinbase.com",
				typeID: "intent",
				versionMajor: 1,
				versionMinor: 0
			)
		case .actions:
			ContentTypeID(
				authorityID: "coinbase.com",
				typeID: "actions",
				versionMajor: 1,
				versionMinor: 0
			)
		case .deletedMessage:
			ContentTypeDeletedMessage
		}
	}

	private func mapReaction(_ reactionPayload: FfiReactionPayload) -> Reaction {
		Reaction(
			reference: reactionPayload.reference,
			action: reactionPayload.action == .added ? .added : .removed,
			content: reactionPayload.content,
			schema: mapReactionSchema(reactionPayload.schema),
			referenceInboxId: reactionPayload.referenceInboxId
		)
	}

	private func mapReactionSchema(_ schema: FfiReactionSchema) -> ReactionSchema {
		switch schema {
		case .unicode:
			.unicode
		case .shortcode:
			.shortcode
		case .custom:
			.custom
		case .unknown:
			.unknown
		}
	}

	private func mapRemoteAttachmentScheme(_ scheme: String) -> RemoteAttachment.Scheme {
		RemoteAttachment.Scheme(rawValue: scheme) ?? .https
	}

	private func mapAttachment(_ ffiAttachment: FfiAttachment) -> Attachment {
		Attachment(
			filename: ffiAttachment.filename ?? "",
			mimeType: ffiAttachment.mimeType,
			data: ffiAttachment.content
		)
	}

	private func mapRemoteAttachment(_ ffiAttachment: FfiRemoteAttachment) throws -> RemoteAttachment {
		try RemoteAttachment(
			url: ffiAttachment.url,
			contentDigest: ffiAttachment.contentDigest,
			secret: ffiAttachment.secret,
			salt: ffiAttachment.salt,
			nonce: ffiAttachment.nonce,
			scheme: mapRemoteAttachmentScheme(ffiAttachment.scheme),
			contentLength: ffiAttachment.contentLength.map { Int($0) },
			filename: ffiAttachment.filename
		)
	}

	private func mapMultiRemoteAttachment(_ multiAttachment: FfiMultiRemoteAttachment) -> MultiRemoteAttachment {
		MultiRemoteAttachment(
			remoteAttachments: multiAttachment.attachments.map { info in
				MultiRemoteAttachment.RemoteAttachmentInfo(
					url: info.url,
					filename: info.filename ?? "",
					contentLength: info.contentLength ?? 0,
					contentDigest: info.contentDigest,
					nonce: info.nonce,
					scheme: info.scheme,
					salt: info.salt,
					secret: info.secret
				)
			}
		)
	}

	private func mapTransactionReference(_ txRef: FfiTransactionReference) -> TransactionReference {
		TransactionReference(
			namespace: txRef.namespace,
			networkId: txRef.networkId,
			reference: txRef.reference,
			metadata: txRef.metadata.map { metadata in
				TransactionReference.Metadata(
					transactionType: metadata.transactionType,
					currency: metadata.currency,
					amount: metadata.amount,
					decimals: metadata.decimals,
					fromAddress: metadata.fromAddress,
					toAddress: metadata.toAddress
				)
			}
		)
	}

	private func mapGroupUpdated(_ groupUpdated: FfiGroupUpdated) -> GroupUpdated {
		var updated = GroupUpdated()
		updated.initiatedByInboxID = groupUpdated.initiatedByInboxId
		updated.addedInboxes = groupUpdated.addedInboxes.map { inbox in
			var inboxEntry = GroupUpdated.Inbox()
			inboxEntry.inboxID = inbox.inboxId
			return inboxEntry
		}
		updated.removedInboxes = groupUpdated.removedInboxes.map { inbox in
			var inboxEntry = GroupUpdated.Inbox()
			inboxEntry.inboxID = inbox.inboxId
			return inboxEntry
		}
		updated.leftInboxes = groupUpdated.leftInboxes.map { inbox in
			var inboxEntry = GroupUpdated.Inbox()
			inboxEntry.inboxID = inbox.inboxId
			return inboxEntry
		}
		updated.metadataFieldChanges = groupUpdated.metadataFieldChanges.map { change in
			var fieldChange = GroupUpdated.MetadataFieldChange()
			fieldChange.fieldName = change.fieldName
			fieldChange.oldValue = change.oldValue ?? ""
			fieldChange.newValue = change.newValue ?? ""
			return fieldChange
		}
		return updated
	}

	private func mapFfiEncodedContent(_ ffiContent: FfiEncodedContent) throws -> EncodedContent {
		var encoded = EncodedContent()
		if let typeId = ffiContent.typeId {
			encoded.type = ContentTypeID(
				authorityID: typeId.authorityId,
				typeID: typeId.typeId,
				versionMajor: Int(typeId.versionMajor),
				versionMinor: Int(typeId.versionMinor)
			)
		}
		encoded.parameters = ffiContent.parameters
		encoded.fallback = ffiContent.fallback ?? ""
		encoded.content = ffiContent.content
		if let compression = ffiContent.compression {
			// Map Int32 compression values to the enum
			// Assuming 0 = deflate, 1 = gzip based on the proto definition
			switch compression {
			case 0:
				encoded.compression = .deflate
			case 1:
				encoded.compression = .gzip
			default:
				// Unknown compression value, leave it unset
				break
			}
		}
		return encoded
	}
}
